utils/e2e-tests/ts/tests/eth/contractsState.ts

```diff
@@ -0,0 +1,63 @@
+import { describe, it, expect, assert } from "vitest";
+import { RunNodeState, runNode } from "../../lib/node";
+import * as eth from "../../lib/ethViem";
+import deposit from "../../lib/abis/deposit";
+import "../../lib/expect";
+import { beforeEachWithCleanup } from "../../lib/lifecycle";
+
+describe("contract account's state", () => {
+  let node: RunNodeState;
+  let publicClient: eth.PublicClientWebSocket;
+  let devClients: eth.DevClientsWebSocket;
+  beforeEachWithCleanup(async (cleanup) => {
+    node = runNode({ args: ["--dev", "--tmp"] }, cleanup.push);
+
+    await node.waitForBoot;
+
+    publicClient = eth.publicClientFromNodeWebSocket(node, cleanup.push);
+    devClients = eth.devClientsFromNodeWebSocket(node, cleanup.push);
+  }, 60 * 1000);
+
+  // See also: https://github.com/humanode-network/humanode/issues/1402
+  it("doesn't vanish after emptying its wallet", async () => {
+    const [alice, _] = devClients;
+
+    const deployContractTxHash = await alice.deployContract({
+      abi: deposit.abi,
+      bytecode: deposit.bytecode,
+      value: 1n, // Even the smallest deposit is enough
+      gas: 150_274n,
+    });
+    const deployContractTxReceipt =
+      await publicClient.waitForTransactionReceipt({
+        hash: deployContractTxHash,
+        timeout: 18_000,
+      });
+    expect(deployContractTxReceipt.status).toBe("success");
+    const contract = deployContractTxReceipt.contractAddress;
+    assert(contract);
+
+    // The nonce of the contract account immediately after creation.
+    const INITIAL_NONCE = 1;
+
+    // If there's a bug in the EVM execution, it will clear the contract account state (including nonce) after `withdrawAll`.
+    const withdrawalTx = await alice.writeContract({
+      address: contract,
+      abi: deposit.abi,
+      functionName: "withdrawAll",
+      gas: 30_585n,
+    });
+    const withdrawalReceipt = await publicClient.waitForTransactionReceipt({
+      hash: withdrawalTx,
+      timeout: 18_000,
+    });
+    expect(withdrawalReceipt.status, "status of withdrawal").toBe("success");
+
+    // Ethereum RPC returns the account nonce in the `eth_getTransactionCount` RPC call.
+    // https://github.com/humanode-network/frontier/blob/locked/polkadot-v0.9.42-2024-09-12/client/rpc/src/eth/state.rs#L116-L169
```

Comment:
What about using the link to frontier master code ?

Reply:
Actually better to link to a specific commit - that way we can ref the exact code

Reply:
Specified https://github.com/humanode-network/frontier/blob/1afab28e8d5aebe7d44f9043b3ba19e9555123dc/client/rpc/src/eth/state.rs#L116-L169

Reply:
See https://eips.ethereum.org/EIPS/eip-7702
This is a bit more complicated since EOAs could have code, too, and EIP 7702 specifies one way of doing it.

Not saying this enum is no good already - didn't think through yet, but just giving you guys a pointer since I was this discussion.